File: src/104. Pandigital Fibonacci ends/104.py
import math

# assumes n is an integer with at most 9 digits
def pandigital(n):
    if n < 10 ** 8:
        return False
    digits = set()
    for i in range(9):
        n, d = divmod(n, 10)
        if d == 0:
            return False
        if d in digits:
            return False
        else:
            digits.add(d)
    return True

def begins_pandigital(n):
    return pandigital(int(str(n)[0:9]))

def ends_pandigital(n):
    return pandigital(n % 10 ** 10)

k = 2
fa, fb = 1, 1
while True:
    fa, fb = fb, fa + fb
    k += 1
    if ends_pandigital(fb) and begins_pandigital(fb):
        print(k)
        break

File: src/104. Pandigital Fibonacci ends/104.rb
require 'set'

class Integer
  def pandigital?
    return false if self < 100_000_000
    digits = Set.new
    n = self
    9.times do
      n, d = n.divmod(10)
      return false if d == 0
      if digits.include? d
        return false
      else
        digits << d
      end
    end
    true
  end

  def begins_pandigital?
    self.to_s[0, 9].to_i.pandigital?
  end

  def ends_pandigital?
    (self % 1_000_000_000).pandigital?    
  end
end

k = 2
fa, fb = 1, 1
while true
  fa, fb = fb, fa + fb
  k += 1
  if fb.ends_pandigital? and fb.begins_pandigital?
    puts k
    break
  end
end